File: network/src/main/kotlin/org/bitlap/network/core/Session.kt
package org.bitlap.network.core

import org.bitlap.common.BitlapConf
import org.bitlap.network.core.operation.OperationHandle
import org.bitlap.network.core.operation.OperationManager
import java.util.concurrent.atomic.AtomicBoolean

/**
 *
 * @author 梦境迷离
 * @since 2021/6/6
 * @version 1.0
 */
interface Session {

    val sessionState: AtomicBoolean
    val sessionHandle: SessionHandle
    val password: String
    val username: String
    val creationTime: Long
    val sessionConf: BitlapConf
    val sessionManager: SessionManager

    var lastAccessTime: Long
    var operationManager: OperationManager

    /**
     * open Session
     * @param sessionConfMap
     * @return SessionHandle The Session handle
     */
    fun open(sessionConfMap: Map<String, String>?): SessionHandle

    /**
     * execute statement
     * @param statement
     * @param confOverlay
     * @return OperationHandle The Operate handle
     */
    fun executeStatement(
        sessionHandle: SessionHandle,
        statement: String,
        confOverlay: Map<String, String>?
    ): OperationHandle

    /**
     * execute statement
     * @param statement
     * @param confOverlay
     * @param queryTimeout
     * @return OperationHandle The Operate handle
     */
    fun executeStatement(
        sessionHandle: SessionHandle,
        statement: String,
        confOverlay: Map<String, String>?,
        queryTimeout: Long
    ): OperationHandle

    fun fetchResults(operationHandle: OperationHandle): RowSet
    fun getResultSetMetadata(operationHandle: OperationHandle): TableSchema
    // TODO: close OperationHandle

    /**
     * close Session
     */
    fun close()
}

File: bitlap-server/src/main/scala/org/bitlap/server/rpc/Session.scala
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.server.rpc

import org.bitlap.common.BitlapConf
import org.bitlap.network.handles._
import org.bitlap.network.models._

import java.util.concurrent.atomic.AtomicBoolean

/** bitlap 会话
 *
 *  @author
 *    梦境迷离
 *  @since 2021/6/6
 *  @version 1.0
 */
trait Session {

  var sessionHandle: SessionHandle
  var password: String
  var username: String
  var sessionConf: BitlapConf
  var sessionManager: SessionManager
  var lastAccessTime: Long
  var operationManager: OperationManager

  def sessionState: AtomicBoolean

  def creationTime: Long

  /** open Session
   *
   *  @param sessionConfMap
   *  @return
   *    SessionHandle The Session handle
   */
  def open(sessionConfMap: Map[String, String] = Map.empty): SessionHandle

  /** execute statement
   *
   *  @param statement
   *  @param confOverlay
   *  @return
   *    OperationHandle The Operate handle
   */
  def executeStatement(
    sessionHandle: SessionHandle,
    statement: String,
    confOverlay: Map[String, String]
  ): OperationHandle

  /** execute statement
   *
   *  @param statement
   *  @param confOverlay
   *  @param queryTimeout
   *  @return
   *    OperationHandle The Operate handle
   */
  def executeStatement(
    sessionHandle: SessionHandle,
    statement: String,
    confOverlay: Map[String, String] = Map.empty,
    queryTimeout: Long
  ): OperationHandle

  def fetchResults(operationHandle: OperationHandle): RowSet

  def getResultSetMetadata(operationHandle: OperationHandle): TableSchema

  // TODO: close OperationHandle

  /** close Session
   */
  def close()
}
